### PR TITLE
Remove `requiresLeadingNewline` from basic format

### PR DESCRIPTION
**Pull request by kimdv**
As part of this comment I started by removing the first part

https://github.com/apple/swift-syntax/pull/1623#discussion_r1196921113

### CODE REVIEW

**Comment by kimdv** [comment]
@swift-ci please test

**Review by ahoppen — APPROVED** [review]
Whooo 🎉

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by kimdv** [comment]
@swift-ci please test Windows

**Comment by kimdv** [comment]
@swift-ci please test macOS

**Comment by kimdv** [comment]
@swift-ci please test macOS

**Comment by ahoppen** [comment]
I prefer `precedencegroup MyPrecedence {}` over `precedencegroup MyPrecedence {\n}`, so updating the tests seems good to me.

To fix the test failures for `SourceKit/Macros/diags.swift`, you need to modify the following file https://github.com/apple/swift/blob/main/test/SourceKit/Macros/diags.swift.response. The test that is being run is in the compiler repo as well: https://github.com/apple/swift/blob/main/test/SourceKit/Macros/diags.swift

For `Macros/macro_expand.swift`, I’m not sure because I don’t see the failure anymore but you most likely need to modify some of the `CHECK` lines in https://github.com/apple/swift/blob/main/test/Macros/macro_expand.swift

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by kimdv** [comment]
@ahoppen it looks like if we want `{}` there was more changes.

**Comment by kimdv** [comment]
@swift-ci please test

File: Sources/SwiftBasicFormat/BasicFormat.swift [hunk 1]
```diff
@@ -178,11 +174,15 @@ open class BasicFormat: SyntaxRewriter {
     }
 
     switch (first?.tokenKind, second?.tokenKind) {
+    case (.leftBrace, .rightBrace):  // empty declarations like `struct {}`.
+      return false
```
Comment (ahoppen) [inline Sources/SwiftBasicFormat/BasicFormat.swift:178]: I would prefer this to be a whitelist that always returns `true` if something matches and only falls through to `return false`. In my experience those are a lot easier to reason about than if you interleave white and blacklists.

So, I’d prefer this to change the below to something like here

https://github.com/apple/swift-syntax/blob/3ed9dcf27c69b991d3cd8eadd64764638f3b0dec/Sources/SwiftBasicFormat/BasicFormat.swift#L260


---

When seeing the examples in the test case below, I think I changed my mind that I prefer to add a newline in between `{` and `}`. I prefer to write `func foo() {}` if I’m deliberately declaring an empty function but in the cases below, we are mostly adding `{` and `}` for Fix-Its and in that case you almost always want to add something between the braces.

Comment (kimdv) [reply Sources/SwiftBasicFormat/BasicFormat.swift:178]: I will revert the change and update the two previously failing test cases

File: Sources/SwiftBasicFormat/BasicFormat.swift [hunk 2]
```diff
@@ -178,11 +174,15 @@ open class BasicFormat: SyntaxRewriter {
...
     case (.multilineStringQuote, .backslash),  // string interpolation segment inside a multi-line string literal
       (.multilineStringQuote, .multilineStringQuote),  // empty multi-line string literal
       (.multilineStringQuote, .stringSegment),  // segment starting a multi-line string literal
       (.stringSegment, .multilineStringQuote),  // ending a multi-line string literal that has a string interpolation segment at its end
-      (.rightParen, .multilineStringQuote):  // ending a multi-line string literal that has a string interpolation segment at its end
+      (.rightParen, .multilineStringQuote),  // ending a multi-line string literal that has a string interpolation segment at its end
+      (_, .rightBrace),
+      (_, .poundEndifKeyword):
```
[the remark this answers is not in the document]
Comment (kimdv) [reply Sources/SwiftBasicFormat/BasicFormat.swift:183]: You are totally right! 
I've added also test case for `#elseif`.

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by kimdv** [comment]
@swift-ci please test

**Review by ahoppen — APPROVED** [review]
Nice to see yet another use of CodeGeneration gone

**Comment by kimdv** [comment]
https://github.com/apple/swift/pull/66106

@swift-ci plese test

**Comment by kimdv** [comment]
https://github.com/apple/swift/pull/66106

@swift-ci please test

**Comment by kimdv** [comment]
https://github.com/apple/swift/pull/66106

@swift-ci please test

**Comment by kimdv** [comment]
@swift-ci please test windows

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by kimdv** [comment]
https://github.com/apple/swift/pull/66106

@swift-ci please test

**Comment by kimdv** [comment]
https://github.com/apple/swift/pull/66106

@swift-ci please test

**Comment by kimdv** [comment]
@ahoppen after some tried I got it working. 

Will you also review: https://github.com/apple/swift/pull/66106

**Pull request merged**